src/lib.rs

```diff
@@ -1426,89 +1398,48 @@ impl EGraph {
-        stop: CompilerPassStop,
-    ) -> Result<Vec<NormCommand>, Error> {
-        let mut result = vec![];
-
-        for command in program {
-            match command {
-                Command::Push(num) => {
-                    for _ in 0..num {
-                        self.push();
-                    }
-                }
-                Command::Pop(num) => {
-                    for _ in 0..num {
-                        self.pop()
-                            .expect("Failed to desugar, popped too many times");
-                    }
-                }
-                _ => {}
-            }
-            result.extend(self.process_command(command, stop)?);
-        }
-        Ok(result)
-    }
-
     pub fn set_underscores_for_desugaring(&mut self, underscores: usize) {
         self.desugar.number_underscores = underscores;
     }
 
-    fn process_command(
-        &mut self,
-        command: Command,
-        stop: CompilerPassStop,
-    ) -> Result<Vec<NormCommand>, Error> {
-        let mut program =
+    fn process_command(&mut self, command: Command) -> Result<Vec<ResolvedNCommand>, Error> {
+        let program =
             self.desugar
                 .desugar_program(vec![command], self.test_proofs, self.seminaive)?;
-        if stop == CompilerPassStop::Desugar {
-            return Ok(program);
-        }
 
-        let type_info_before = self.type_info().clone();
-
-        self.desugar.type_info.typecheck_program(&program)?;
-        if stop == CompilerPassStop::TypecheckDesugared {
-            return Ok(program);
-        }
-
-        // now add term encoding
-        if self.terms_enabled {
-            let program_terms = TermState::add_term_encoding(self, program);
-            program = self.desugar.desugar_program(program_terms, false, false)?;
-        }
-
-        if stop == CompilerPassStop::TermEncoding {
-            return Ok(program);
-        }
-
-        // reset type info
-        self.desugar.type_info = type_info_before;
-        self.desugar.type_info.typecheck_program(&program)?;
-        if stop == CompilerPassStop::TypecheckTermEncoding {
-            return Ok(program);
-        }
+        let program = self.type_info_mut().typecheck_program(&program)?;
 
         Ok(program)
     }
 
+    /// Run a program, represented as an AST.
+    /// Return a list of messages.
     pub fn run_program(&mut self, program: Vec<Command>) -> Result<Vec<String>, Error> {
-        let should_run = true;
-
         for command in program {
             // Important to process each command individually
             // because push and pop create new scopes
-            for processed in self.process_command(command, CompilerPassStop::All)? {
-                self.run_command(processed, should_run)?;
+            for processed in self.process_command(command)? {
+                if self.run_mode.show_egglog() {
+                    // This is a hack: we have to do this because process_commands are indeed
```

Comment:
Having trouble following what's happening here.

My read here is that `show_egglog()` is essentially a "dry run" mode where the only commands we want to run are the ones that we need to make everything well-scoped (Push/Pop). 

But then how does that relate to typechecking a function twice? When would we do this twice? And aren't we still typechecking in `process_command?` ?

Reply:
You are right. I was specifically talking about "typechecking a function twice" because, in our test, we have code like
```
(sort Expr)
(push)
(function f () Expr)
(pop)
(push)
(function f () Expr)
(pop)
```
It is only when `(push)` and `(pop)` are run that this does not lead to an error. But I agree that "typechecking twice" is not fundamental to the issue here. I'll update the comment to reflect this.